[Tables] Migration Guide (#15857)
Azure/azure-sdk-for-js: sdk/tables/data-tables/MigrationGuide.md

# Guide for migrating to `@azure/data-tables` from `azure-storage`

This guide is intended to assist in the migration to `@azure/data-tables` from the legacy `azure-storage` package. It will focus on side-by-side comparisons for similar operations between the two packages.

We assume that you are familiar with `azure-storage`. If you are new to the Azure Tables client library for JavaScript, please refer to the [README](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/tables/data-tables/README.md) and [samples](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/tables/data-tables/samples/v12) rather than this guide.

> ## Table of contents

Reviewer: I think GitHub will auto-generate the TOC

### Cross Service SDK improvements

The modern `@azure/data-tables` client library also provides the ability to share in some of the cross-service improvements made to the Azure development experience, such as

- Using the new `@azure/identity` library to share a single authentication approach between clients
- A unified logging and diagnostics pipeline offering a common view of the activities across each of the client libraries
- Use of promises rather than callbacks for a simplified programming experience
- Use of async iterators in paging APIs

## Important changes

### Package name

The modern client library is named `@azure/data-tables` and was released beginning with version 12. The legacy client library is named `azure-storage` with version of 2.x.x or below.

### Constructing the clients

> Previously in `azure-storage`, you would use `createTableService` which can be used to get an instance of the `TableService` in order to perform service level operations

Reviewer: Missing a period at the end of the sentence (same with a few other places)
```suggestion
Previously in `azure-storage`, you would use `createTableService` which can be used to get an instance of the `TableService` in order to perform service level operations.
```

```javascript
const azure = require('azure-storage');
var tableService = azure.createTableService("<connection-string>");
```

Now, in `@azure/data-tables`, we need a TableServiceClient for service level operations.
```javascript
const { TableServiceClient } = require('@azure/data-tables');
const tableService = TableServiceClient.fromConnectionString("<connection-string>")
```

### Creating a Table
Previously in `azure-storage`, we'd use a `TableService` instance to create a table. The `createTable` method would take a callback to execute once the table has been created. This forces sequential operations to be inside the callback, potentially creating a callback chain

```javascript 
const azure = require('azure-storage');
var tableService = azure.createTableService("<connection-string>");

const tableName = "<table-name>"
tableService.createTable(tableName);

```

> With `@azure/data-tables` we have access to all table level operations directly from the `TableServiceClient`. Because the table service client is not affinitized to any one table, it is ideal for scenarios where you need to create, delete, or list more than one Table.

Reviewer: ```suggestion
With `@azure/data-tables` we have access to all table level operations directly from the `TableServiceClient`. Because the table service client is not affinitized to any one table, it is ideal for scenarios where you need to create, delete, or list more than one table.
```

}

await tableClient.createEntity(task1);
```

### Fetching a single entity from the table
Both clients allow for fetching a single entity from the table if the partitionKey and rowKey are known.

Previously in `azure-storage`, to access the entity we'd need to do it inside the callback on `retrieveEntity`

```javascript 
const azure = require('azure-storage');
var tableService = azure.createTableService("<connection-string>");

const tableName = "<table-name>"
tableService.retrieveEntity(tableName, "hometasks", "1", function(error, result, response) {
    if (!error) {
        // result contains the entity
        console.log(result);
    }
})
```

Now with `@azure/data-tables` we use `getEntity` on the TableClient, the return type is a Promise of the entity which can be awaited, making the code cleaner.

```javascript 
const { TableClient, AzureNamedKeyCredential } = require('@azure/data-tables');
const tableName = "<table-name>"
const tableClient = new TableClient(tablesEndpoint, tableName, new AzureNamedKeyCredential("<accountName>","<accountKey>"));

const entity = await tableClient.getEntity("hometasks", "1");
console.log(entity);
```

### Querying data from the table

> Previously in `azure-storage`, querying a table looked as follows, and didn't provide a built in way to handle pagination

Reviewer: ```suggestion
Previously in `azure-storage`, querying a table didn't provide a built in way to handle pagination, looking as follows.
```

```javascript
const azure = require('azure-storage');
const tableService = azure.createTableService("<connection-string>");
const tableName = "<table-name>"
const query = new azure.TableQuery()
  .where('PartitionKey eq ?', 'part2');

let entities = [];

function listEntities(query, continuationToken, callback) {
    tableService.queryEntities(tableName, query, function(error, result){
        entities.push(result.entities);
        const token = result.continuationToken;
        if(token) {
            listEntities(query, continuationToken, callback)
        } else {
            console.log("completed getting all entities");
            callback();
        }
    })
}

listEntities(query, null, function() {
    console.log(entities);
})

```

> In the new `@azure/data-tables` we return a `PagedAsyncIterableIterator` that handles fetching all the pages internally, and can be iterated easily.

Reviewer: ```suggestion
In the new `@azure/data-tables` we return a `PagedAsyncIterableIterator` that handles the details of pagination internally, simplifying the task of iteration.
```

  dueDate: new Date(2015, 6, 20)
};
const task2 = {
  PartitionKey: "hometasks",
  RowKey: "2",
  description: "Wash the dishes",
  dueDate: new Date(2015, 6, 20)
};

const tableActions = [["create", task1], ["create", task2]];

await tableClient.submitTransaction(tableActions);
console.log("Transaction completed")
```

For convenience also provide a helper that works in a similar way as `TableBatch` in `azure-storage`.

```javascript
const { TableClient, AzureNamedKeyCredential, TableTransaction } = require('@azure/data-tables');
const tableName = "<table-name>"
const tableClient = new TableClient(tablesEndpoint, tableName, new AzureNamedKeyCredential("<accountName>","<accountKey>"));

const task1 = {
  PartitionKey: "hometasks",
  RowKey: "1",
  description: "Take out the trash",
  dueDate: new Date(2015, 6, 20)
};
const task2 = {
  PartitionKey: "hometasks",
  RowKey: "2",
  description: "Wash the dishes",
  dueDate: new Date(2015, 6, 20)
};

const transaction = new TableTransaction();
transaction.createEntity(task1);
transaction.createEntity(task2);

await tableClient.submitTransaction(transaction.actions);
console.log("Transaction completed")
```

### Sequential Actions

> Previously in `azure-storage`, all the operations took a callback used to execute actions in sequence. For example, to create a table and then insert two entities

Reviewer: This sentence seems to have ended too early. Did you intend to have more after "insert two entities..."?

```javascript 
const azure = require('azure-storage');
var tableService = azure.createTableService("<connection-string>");

const tableName = "<table-name>"
tableService.createTable(tableName, function() {
    tableService.insertEntity(tableName, {PartitionKey: "p1", RowKey: "r1", foo: "bar"}, function() {
        tableService.insertEntity(tableName, {PartitionKey: "p2", RowKey: "r2", foo: "baz"});
    })
})
```

With `@azure/data-tables` we work with promises which makes the programming experience better

```javascript 
const { TableClient, AzureNamedKeyCredential } = require('@azure/data-tables');
const tableName = "<table-name>"
const tableClient = new TableClient(tablesEndpoint, tableName, new AzureNamedKeyCredential("<accountName>","<accountKey>"));

await tableClient.createTable();
await tableClient.createEntity({partitionKey: "p1", rowKey: "r1", foo: "bar"});
await tableClient.createEntity({partitionKey: "p2", rowKey: "r2", foo: "baz"});

```

## Additional samples

> More examples can be found at [here](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/tables/data-tables/samples/v12)

Reviewer: ```suggestion
More samples can be found [here](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/tables/data-tables/samples/v12)
```

To improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) was created for all languages to drive a consistent experience with established API patterns for all services. A set of [TypeScript & JavaScript Guidelines](https://azure.github.io/azure-sdk/typescript_introduction.html) was also introduced to ensure that TypeScript clients have a natural and idiomatic feel with respect to the TypeScript and JavaScript ecosystems. The new `@azure/data-tables` follows these guidelines.

### Cross Service SDK improvements

The modern `@azure/data-tables` client library also provides the ability to share in some of the cross-service improvements made to the Azure development experience, such as

> - Using the new `@azure/identity` library to share a single authentication approach between clients

Author: I have a PR adding support for AAD on Storage endpoints :P but yeah that's new and will not be needed for migration. I'll remove this reference

- A unified logging and diagnostics pipeline offering a common view of the activities across each of the client libraries
- Use of promises rather than callbacks for a simplified programming experience
- Use of async iterators in paging APIs

## Important changes

### Package name

The modern client library is named `@azure/data-tables` and was released beginning with version 12. The legacy client library is named `azure-storage` with version of 2.x.x or below.

### Constructing the clients

Previously in `azure-storage`, you would use `createTableService` which can be used to get an instance of the `TableService` in order to perform service level operations.

```javascript
const azure = require("azure-storage");
const tableService = azure.createTableService("<connection-string>");
```

Now, in `@azure/data-tables`, we need a TableServiceClient for service level operations.

```javascript
const { TableServiceClient } = require("@azure/data-tables");

> const tableService = TableServiceClient.fromConnectionString("<connection-string>");

Author: I think the main reason is that it is hard hard to disambiguate with the other constructor overloads (last one is a collision)
![image](https://user-images.githubusercontent.com/21109913/122819986-f820fd80-d28f-11eb-9113-0736cf37a918.png)

Plus storage has this model as well

```

### Creating a Table

> Previously in `azure-storage`, we'd use a `TableService` instance to create a table. The `createTable` method would take a callback to execute once the table has been created. This forces sequential operations to be inside the callback, potentially creating a callback chain

Reviewer 2: ```suggestion
Previously in `azure-storage`, you would use a `TableService` instance to create a table. The `createTable` method would take a callback to execute once the table has been created. This forces sequential operations to be inside the callback, potentially creating a callback chain
```

```javascript
const azure = require("azure-storage");
const tableService = azure.createTableService("<connection-string>");

const tableName = "<table-name>";
tableService.createTable(tableName, function() {
  console.log(`Table created`);
});
```

> With `@azure/data-tables` we have access to all table level operations directly from the `TableServiceClient`. Because the table service client is not affinitized to any one table, it is ideal for scenarios where you need to create, delete, or list more than one table.

Reviewer 2: ```suggestion
With `@azure/data-tables` you have access to all table level operations directly from the `TableServiceClient`. Because the table service client is not affinitized to any one table, it is ideal for scenarios where you need to create, delete, or list more than one table.
```

```javascript
const azure = require("azure-storage");
const tableName = "<table-name>";
const tableService = azure.createTableService("<connection-string>");

const task1 = {
  PartitionKey: { _: "hometasks" },
  RowKey: { _: "1" },
  description: { _: "take out the trash" },
  dueDate: { _: new Date(2015, 6, 20), $: "Edm.DateTime" }
};

tableService.insertEntity(tableName, task1, function() {
  console.log("Entity inserted");
});
```

The other way in `azure-storage` to insert an entity was to use the entityGenerator which helped abstracting the creation of the value object described above

```javascript
const azure = require("azure-storage");
const entGen = azure.TableUtilities.entityGenerator;
const tableName = "<table-name>";
const tableService = azure.createTableService("<connection-string>");

const task1 = {
  PartitionKey: entGen.String("hometasks"),
  RowKey: entGen.String("1"),
  description: entGen.String("take out the trash"),
  dueDate: entGen.DateTime(new Date(2015, 6, 20))
};

tableService.insertEntity(tableName, task1, function() {
  console.log("Entity inserted");
});
```

> Now in the new `@azure/data-tables` SDK, in order to have more idiomatic property names in our entities we have moved to `partitionKey` and `rowKey` (camel case). Also you no longer need to follow the value object structure or entityGenerator anymore.

Reviewer 2: ```suggestion
Now in the new `@azure/data-tables` SDK, in order to have more idiomatic property names in our entities we have moved to `partitionKey` and `rowKey` (camel case). Also you no longer need to use the value object structure or entityGenerator anymore, instead use normal JavaScript values.
```

```javascript
const { TableClient, AzureNamedKeyCredential } = require("@azure/data-tables");
const tableName = "<table-name>";
const tablesEndpoint = "https://<account-name>.table.core.windows.net";

const tableClient = new TableClient(
  tablesEndpoint,
  tableName,
  new AzureNamedKeyCredential("<accountName>", "<accountKey>")
);

// Creates the table with `tableName` if it doesn't exist
const task1 = {
  partitionKey: "hometasks",
  rowKey: "1",
  description: "take out the trash",
  dueDate: new Date(2015, 6, 20)
};

await tableClient.createEntity(task1);
```

> If you are using TypeScript, `@azure/data-tables` package provides a type that can help you build your entities providing static check to make sure the required `rowKey` and `partitionKey` properties are present.

Reviewer 2: ```suggestion
If you are using TypeScript, the `@azure/data-tables` package provides a type, `TableEntity`, that can help you build your entities providing static check to make sure the required `rowKey` and `partitionKey` properties are present.
```
